### ISSUE TITLE
Unnecessary cast false positive in expression trees

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#58887]
**Version Used**: 

main

**Steps to Reproduce**:

[SharpLab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQBkCWAdgI7YCiAHgA4BOApgM4P4D2hDA3MslgGyYAWdAFkAFFlQBtALroAhjQDmDAJTIA3snTb0AYwAWCgPzoK6ALzpCAVwA2trki07KtRszYAeKAFZPAIxYWWwA+EPQAF0YIuAt0URULcLNzSwByODTHHXRXeiZWQm8/QOCwyOiYOISk0wtLUQMFFQys520890LigKDQ8KiGCIBmasTzZPr4ppojFsyOdAB6JfQAJToAWxYANyJFSP18Bj05Ib1DQkVGdH86Qz2WGmQAXyA==)

```csharp
using System;
using System.Linq.Expressions;

static void M(string[] args)
{
    char? x = null;

    Expression<Func<bool>> test1 = () => x == '1';
    Expression<Func<bool>> test2 = () => x == (char)'1';
    Expression<Func<bool>> test3 = () => x == (char?)'1'; // Removing this cast changes behavior
}
```

@CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
Changes behavior in what sense?

**Comment by CyrusNajmabadi** [comment]
Also, i'm really struggling to see how this coudl ever be relevant in a real world case.  Absent a true example of this being a problem somewhere, i'm inclined to not fix.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi An example of behavioral change:

```csharp
using System;
using System.Linq.Expressions;

char? x = null;

printRightOperandType(() => x == (char?)'1');
printRightOperandType(() => x == '1');


void printRightOperandType(Expression<Func<bool>> expression)
{
    var binaryExpr = (BinaryExpression)expression.Body;
    var right = (UnaryExpression)binaryExpr.Right;
    Console.WriteLine(right.Operand.Type);
}
```

I think this might be relevant in LINQ to SQL scenarios, but I'm not sure.

**Comment by CyrusNajmabadi** [comment]
Yeah.  My caring level is near zero.  For pedagogical reasons though, @jaredpar do you know why this behaves any different?  

With a normal lambda all the three forms are rewritten to the exact same thing afaict.  I'm curious why the cast is removed in the second expr-tree case, but not the 3rd.

--

Note: @Youssef1313 IMO, if your code actually depends on this, that's a sign that somethign is very smelling with said API.  I'm ok leaving the behavior for now and saying that one shoudl suppress the analyzer in that case.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi The seemingly related scenario is https://github.com/dotnet/roslyn/issues/58886 (and the issue I linked there). Those two issues were the reason I opened this. On compiler side, this is pretty much by design. Expression trees are very sensitive to code changes, and there are usually reports for false positives in it for different analyzers.

For this specific case, I think we should detect if we're in an expression tree, and then report unnecessary cast if and only if the "source" and "destination" types are **exactly** the same.

**Comment by CyrusNajmabadi** [comment]
Enh.  I'm really fine with this. From the language perspective these are all truly equal. If your use case it's such that you must have the `char?` cast to work (but somehow we're not affected by the fact that a `char` cast wound do nothing), then I think that's just life :-)

**Comment by Youssef1313** [comment]
@CyrusNajmabadi So you're more with closing the issue? Or keeping it in backlog for community contribution (under the condition that the fix doesn't add much complexity to the code)?

**Comment by CyrusNajmabadi** [comment]
i'm ok keeping it open, but waiting for a true case to hit where this matters.  e.g. a real API out there where people were using this idiom, and were actually affected by the analyzer.

**Comment by CyrusNajmabadi** [comment]
Closing out due to lack of feedback.
